vertical-pod-autoscaler/recommender/metrics/metrics_client.go

```diff
@@ -0,0 +1,173 @@
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1lister "k8s.io/kubernetes/pkg/client/listers/core/v1"
+	resourceclient "k8s.io/metrics/pkg/client/clientset_generated/clientset/typed/metrics/v1alpha1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/metrics/pkg/apis/metrics/v1alpha1"
+)
+
+type MetricsClient interface {
+	GetContainersUtilization() ([]*ContainerUtilizationSnapshot, error)
+}
+
+type metricsClient struct {
+	metricsGetter   resourceclient.PodMetricsesGetter
+	podLister       v1lister.PodLister
+	namespaceLister v1lister.NamespaceLister
+}
+
+func NewMetricsClient(metricsGetter resourceclient.PodMetricsesGetter, podLister v1lister.PodLister, namespaceLister v1lister.NamespaceLister) MetricsClient {
+	return &metricsClient{
+		metricsGetter:   metricsGetter,
+		podLister:       podLister,
+		namespaceLister: namespaceLister,
+	}
+}
+
+func (client *metricsClient) GetContainersUtilization() ([]*ContainerUtilizationSnapshot, error) {
+
+	usageSnapshots, err := client.getContainersUsage()
+	containerSpecs, err := client.getContainersSpec()
+	if err != nil {
+		return nil, err
+	}
+
+	utilizationSnapshots, err := mergeIntoUtilization(usageSnapshots, containerSpecs)
+	if err != nil {
+		return nil, err
+	}
+	return utilizationSnapshots, nil
+}
+
+func (client *metricsClient) getContainersSpec() ([]*containerSpecification, error) {
+
+	var containerSpecs []*containerSpecification
+
+	pods, err := client.podLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pod := range pods {
+		for _, container := range pod.Spec.Containers {
+			containerSpec := newContainerSpec(container, pod)
+			containerSpecs = append(containerSpecs, containerSpec)
+		}
+	}
+
+	return containerSpecs, nil
+
+}
+
+func (client *metricsClient) getContainersUsage() ([]*containerUsageSnapshot, error) {
+
+	var usageSnapshots []*containerUsageSnapshot
+
+	namespaces, err := client.getAllNamespaces()
+	if err != nil {
+		return nil, err
+	}
+	for _, namespace := range namespaces {
+		podMetricsList, err := client.metricsGetter.PodMetricses(namespace).List(metav1.ListOptions{})
+		if err != nil {
+			return nil, err
+		}
+		for _, podMetrics := range podMetricsList.Items {
+			containerSnapshots := createContainerUsageSnapshots(podMetrics)
+			usageSnapshots = append(usageSnapshots, containerSnapshots...)
+		}
+	}
+
+	return usageSnapshots, nil
+}
+
+func createContainerUsageSnapshots(podMetrics v1alpha1.PodMetrics) []*containerUsageSnapshot {
+	snapshots := make([]*containerUsageSnapshot, len(podMetrics.Containers))
+	for i, containerMetrics := range podMetrics.Containers {
+		snapshots[i] = newContainerUsageSnapshot(containerMetrics, podMetrics)
+	}
+	return snapshots
+}
+
+func mergeIntoUtilization(snapshots []*containerUsageSnapshot, specifications []*containerSpecification) ([]*ContainerUtilizationSnapshot, error) {
```

Comment:
I called it megre because the function is combining/merging data from 2 sources into a single output and there is not much calculation.

Would `calculateUtilization` sounded better?